mozci/push.py

```diff
@@ -294,14 +296,44 @@ def add(result):
         except MissingDataError:
             pass
 
-        # Gather information from the unittest table. We allow missing data for this table because
-        # ActiveData only holds very recent data in it, but we have fallbacks on Taskcluster
-        # artifacts.
-        try:
-            add(run_query("push_tasks_results_from_unittest", args))
-        except MissingDataError:
-            pass
+        # Let's gather error/results from cache or AD/Taskcluster
+        test_tasks_results = adr.config.cache.get(self.push_uuid, {})
+        was_cached = len(test_tasks_results.keys()) != 0
+        if not was_cached:
+            # Gather information from the unittest table. We allow missing data for this table because
+            # ActiveData only holds very recent data in it, but we have fallbacks on Taskcluster
+            # artifacts.
+            try:
+                add(run_query("push_tasks_results_from_unittest", args, cache=False))
+            except MissingDataError:
+                pass
+
+        tasks = Push._normalized_tasks(tasks)
+
+        # Add any data available in the cache
+        if was_cached:
+            # Let's add cached error summaries to TestTasks
+            for t in tasks:
+                if isinstance(t, TestTask):
+                    error_summary = test_tasks_results.get(t.id)
+                    # Only Test tasks have stored error summary information in the cache
+                    if error_summary:
+                        t._errors = error_summary["errors"]
+                        t._results = error_summary["results"]
+            logger.debug(
+                "Fetched tasks errors/results for {} from the cache".format(
+                    self.push_uuid
+                )
+            )
 
+        # Let's gather any new data missing
+        tasks = self.gather_missing_results(tasks)
```

Comment:
Nit: since we have already another function in Push which does almost the same, could you remove this and just use `get_groups_for_tasks`?
```
# Load errors and results for all tasks.
collections.deque(self.get_groups_for_tasks(), maxlen=1)
```

Reply:
Oops, you called get_groups_for_tasks but didn't consume it! I'll fix it in a follow-up commit.

Reply:
The follow up code refered to:
https://github.com/mozilla/mozci/commit/418b20339d14415adbbeee6619dd397a12ccb965